Add custom module guide - Closes #881 (#882)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/app-development/module.adoc

:url_references_schemas: references/schemas.adoc
:url_tutorials_hello: tutorials/hello-world.adoc

This guide explains step-byb-step, how to create a custom module for a blockchain application built with the Lisk SDK.

.Prerequisites
[NOTE]
====
To follow this guide, the following is assumed:

* a basic blockchain application is already initiated, like explained in the previous guide xref:{url_guides_setup}[].
====

First, create a new file named after the new module, for example `my-module.js`.

----
├── blockchain_app
│   ├── index.js
│   ├── my-module.js
│   └── package.json
----

== Creating the module class

Now open `my-module.js` and import the `BaseModule` from the `lisk-sdk` package:

[source,js]
----
const { BaseModule } = require('lisk-sdk');
----

Next, define a new class `MyModule`, which extends from the `BaseModule`:

[source,js]
----
const { BaseModule } = require('lisk-sdk');

class MyModule extends BaseModule {

}

module.exports = MyModule;
----

== Setting name and ID of the module

Inside of the `myModule` class, define the different properties of the module:

`id`(number)::
The ID of the module.
Must be unique in the application.

> The ID is used for example to for creation of transactions of the module.

Author: ```suggestion
The ID is used for example during the creation of transactions of the module.
```

Minimum value is `1024`.
`name`(string)::
The name of the module.
Must be unique in the application.
The name is used for example to subscribe to events, or to invoke actions of the module.

[source,js]
----
const { BaseModule } = require('lisk-sdk');

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
}

module.exports = MyModule;
----

== Defining the account schema

> The account schema defines which propeties are added to user accounts by the module.

Reviewer: ```suggestion
The account schema defines which properties are added to user accounts by the module.
```

TIP: For more information about schemas and how they are used in the Lisk SDK, check out the xref:{url_references_schemas}[].

The new properties will be added under a key that is named after the module name.

> Inside this new key, the module-specific properties for every account are saved.

Reviewer: ```suggestion
The module-specific properties for every account are saved inside this key.
```

    key1 : "",
    key2 : false,
    key3 : 0
  }
}
----

Now include the schema in the module:

[source,js]
----
const { BaseModule } = require('lisk-sdk');
const { myAccountSchema } = require('./schemas.js'); <1>

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
  accountSchema = myAccountSchema; <2>
}

module.exports = MyModule;
----

<1> Require the schema.
<2> Set the `accountSchema` of the module to the imported schema.

== Adding transaction assets to the module

A module can include various custom transaction assets, that provide new transaction types to the application.

Before a new asset can be added, it is first required to create the custom asset as described in the xref:{url_guides_asset}[] guide.

Assuming an asset, `myAsset` has been created for the module, then it can be included as shown below:

[source,js]
----
const { BaseModule } = require('lisk-sdk');
const { myAccountSchema } = require('./schemas.js');
const { MyAsset } = require('./my-asset.js');

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
  accountSchema = myAccountSchema;
  transactionAssets = [ new myAsset() ];
}

module.exports = MyModule;
----

> == Adding an interface by providing reducers, actions and events

Reviewer: ```suggestion
== Adding an interface by providing reducers, actions, and events
```

Author: I don't think a comma here is correct.

Reviewer: Well, normally you would have a comma before 'and' in this type of sentence whereby one is describing 3 separate topics. However, it is certainly not critical. Let's leave it as you had it originally Mona, as this is your guide. (Which by the way looks excellent).

TIP: See the xref:{url_rpc_endpoints}[] page for more information.

`events`::
A list of events this module emits.
Plugins and external services an subscribe to these events with the API client.
`actions`::
A list of actions that plugins and external services can invoke  via the API client.
`reducers`::
A list of actions that other modules of the application can invoke.

[source,js]
----
const { BaseModule } = require('lisk-sdk');
const { myAccountSchema } = require('./schemas.js');
const { myAsset } = require('./my-asset.js');

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
  accountSchema = myAccountSchema;
  transactionAssets = [ new myAsset() ];
  actions = {
    myAction: async () => {
        // Returns some data
    },
    anotherAction: async (params) => {
        // Returns some other data
    }
  };
  events = ['myEvent','anotherEvent'];
  reducers = {
    myReducer: async (params, stateStore) => {
      // Returns some data
    },
    anotherReducer: async (params, stateStore) => {
      // Returns some other data
    }
  };
}

module.exports = MyModule;
----

[TIP]
====

> What events, actions and reducers to create for a module, or if these interfaces are needed at all is individual for every module, as it heavily depends on which functionality the module intends to provide to the application.

Author: ```suggestion
What events, actions and reducers are used within a module, or if these interfaces are actually required at all , will be a specific individual requirement, as it is heavily dependant on which functionality the module intends to provide to the application.
```

= Creating a custom module
:sectnums:
// Project URLS
:url_guides_setup: guides/app-development/setup.adoc
:url_guides_asset: guides/app-development/transaction-asset.adoc
:url_modules_dpos: dpos-module.adoc
:url_rpc_endpoints: rpc-endpoints.adoc
:url_tutorials_hello: tutorials/hello-world.adoc

This guide explains step-byb-step, how to create a custom module for a blockchain application built with the Lisk SDK.

.Prerequisites
[NOTE]
====
To follow this guide, the following is assumed:

* a basic blockchain application is already initiated, like explained in the previous guide xref:{url_guides_setup}[].
====

First, create a new file named after the new module, for example `my-module.js`.

> 

Reviewer 2: seems like there is empty block here?

Author: I added now the file tree

[source,js]
----
const { BaseModule } = require('lisk-sdk');
----

Next, define a new class `MyModule`, which extends from the `BaseModule`:

[source,js]
----
const { BaseModule } = require('lisk-sdk');

class MyModule extends BaseModule {

}

module.exports = MyModule;
----

== Setting name and ID of the module

Inside of the `myModule` class, define the different properties of the module:

`id`(number)::
The ID of the module.
Must be unique in the application.
The ID is used for example to for creation of transactions of the module.
Minimum value is `1024`.
`name`(string)::
The name of the module.
Must be unique in the application.
The name is used for example to subscribe to events, or to invoke actions of the module.

[source,js]
----
const { BaseModule } = require('lisk-sdk');

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
}

module.exports = MyModule;
----

== Defining the account schema

The account schema defines which propeties are added to user accounts by the module.

The new properties will be added under a key that is named after the module name.
Inside this new key, the module-specific properties for every account are saved.

> DataType must be one of `string`, `bytes`, `boolean`, `uint32`, `sint32`, `uint64` or `sint64`.

Reviewer 2: this explanation should be separately explained and link to there? this encoding/decoding is not only related to account

Author: Fair point! Do you have a suggestion where the content should be moved to in the current structure? Or would it make sense to have a dedicated page for explaining schemas in the `references`?

Reviewer 2: I think it would make sense to have dedicated page

Author: I added now a new page to references, and added a link to it in the module guide, please check :)

      },
    },
    key6: {
      type: "array",
      fieldNumber: 6,
      items: {
        dataType: "bytes",
      },
    },
  },
  // Default values for the different properties
  default: {
    key1 : "",
    key2 : false,
    key3 : 0,
    key4 : {},
    key5 : [],
    key6 : [],
  }
}
----

Now include the schema in the module:

[source,js]
----
const { BaseModule } = require('lisk-sdk');
const { myAccountSchema } = require('./schemas.js'); <1>

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
  accountSchema = myAccountSchema; <2>
}

module.exports = MyModule;
----

<1> Require the schema.
<2> Set the `accountSchema` of the module to the imported schema.

== Adding transaction assets to the module

A module can include various custom transaction assets, that provide new transaction types to the application.

Before a new asset can be added, it is first required to create the custom asset as described in the xref:{url_guides_asset}[] guide.

Assuming you have created an asset `myAsset` for the module, then it can be included like this:

[source,js]
----
const { BaseModule } = require('lisk-sdk');
const { myAccountSchema } = require('./schemas.js');

> const { myAsset } = require('./my-asset.js');

Reviewer 2: ```suggestion
const { MyAsset } = require('./my-asset.js');
```

class MyModule extends BaseModule {
  id = 1024;
  name = 'my-module';
  accountSchema = myAccountSchema;
  transactionAssets = [ new myAsset() ];
  actions = {
    myAction: async () => {
        // Returns some data
    },
    anotherAction: async (params) => {
        // Returns some other data
    }
  };
  events = ['myEvent','anotherEvent'];
  reducers = {
    myReducer: async (params, stateStore) => {
      // Returns some data
    },
    anotherReducer: async (params, stateStore) => {
      // Returns some other data
    }
  };
}

module.exports = MyModule;
----

[TIP]
====
What events, actions and reducers to create for a module, or if these interfaces are needed at all is individual for every module, as it heavily depends on which functionality the module intends to provide to the application.

The best way to understand what to provide here is to look at existing examples in the Lisk SDK default modules, or examples of other blockchain applications built with the Lisk SDK, for example the xref:{url_tutorials_hello}[] application.
====

=== Data access for actions

Blockchain data can be accessed in a module via `this._dataAccess`.

> The data access is mainly used in the implementation of the actions to retrieve certain information from the blockchain.

Reviewer 2: ```suggestion
The data access is only used in the implementation of the actions to retrieve certain information from the blockchain.
```

`afterGenesisBlockApply()`::
Code in here is applied after the genesis block is applied.
`beforeBlockApply()`::
Code in here is applied before each block is applied.
`afterBlockApply()`::
Code in here is applied after each block is applied.

=== Lifecycle hooks

.Lifecycle hooks example
[source,js]
----
async beforeTransactionApply({transaction, stateStore, reducerHandler}) {
    // Code in here is applied before each transaction is applied.
};

async afterTransactionApply({transaction, stateStore, reducerHandler}) {
  // Code in here is applied after each transaction is applied.
  if (transaction.moduleID === this.id && transaction.assetID === MyAssetID) {

    const myAsset = codec.decode(
      myAssetSchema,
      transaction.asset
    );

    this._channel.publish('my-module:myEvent', {
      sender: transaction._senderAddress.helloString
    });
  }
};
async afterGenesisBlockApply({genesisBlock, stateStore, reducerHandler}) {
  // Set the hello counter to zero after the genesis block is applied
  await stateStore.chain.set(
    CHAIN_STATE_HELLO_COUNTER,
    codec.encode(helloCounterSchema, { helloCounter: 0 })
  );
};

> async beforeBlockApply(block, stateStore, reducerHandler) {

Reviewer 2: input is object here

    // Code in here is applied before each block is applied.
}

> async afterBlockApply(block, stateStore, reducerHandler, consensus) {

Reviewer 2: saje as above
